File: app/src/main/java/com/suzei/minote/data/NotesDatabase.java
package com.suzei.minote.data;

import android.content.Context;

import com.suzei.minote.data.dao.NotesDao;
import com.suzei.minote.data.entity.Notes;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.migration.Migration;
import androidx.sqlite.db.SupportSQLiteDatabase;

//  If the database updated, Increment version by 1
@Database(entities = Notes.class, version = 3)
public abstract class NotesDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "notes_storage.db";

    private static NotesDatabase sInstance;

    public abstract NotesDao notesDao();

    private static final Object sLock = new Object();

    static final Migration MIGRATION_2_3 = new Migration(2, 3) {

        @Override
        public void migrate(@NonNull SupportSQLiteDatabase database) {
            // Since we didn't alter the table, there's nothing else to do here.
        }

    };

    public static NotesDatabase getInstance(Context context) {
        synchronized (sLock) {
            if (sInstance == null) {
                sInstance = Room.databaseBuilder(
                        context.getApplicationContext(),
                        NotesDatabase.class,
                        DATABASE_NAME)
                        .addMigrations(MIGRATION_2_3)
                        .build();
            }
            return sInstance;
        }
    }

}

File: app/src/main/java/com/suzei/minote/data/NotesDatabase.kt
package com.suzei.minote.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.suzei.minote.data.converter.DateTimeTypeConverter
import com.suzei.minote.data.dao.NotesDao
import com.suzei.minote.data.dao.TodoDao
import com.suzei.minote.data.dao.TodoItemDao
import com.suzei.minote.data.entity.Notes
import com.suzei.minote.data.entity.Todo
import com.suzei.minote.data.entity.TodoItem

//  If the database updated, Increment version by 1
@Database(entities = [Notes::class, Todo::class, TodoItem::class], version = 6)
@TypeConverters(DateTimeTypeConverter::class)
abstract class NotesDatabase : RoomDatabase() {

    abstract fun notesDao(): NotesDao
    abstract fun todoDao(): TodoDao
    abstract fun todoItemDao(): TodoItemDao

    companion object {

        private const val DATABASE_NAME = "notes_storage.db"

        private val sLock = Any()

        fun getInstance(context: Context): NotesDatabase {
            synchronized(sLock) {
                return Room.databaseBuilder(
                        context.applicationContext,
                        NotesDatabase::class.java,
                        DATABASE_NAME)
                        .addMigrations(MIGRATION_2_3, MIGRATION_4_5)
                        .build()
            }
        }

        private val MIGRATION_2_3: Migration = object : Migration(2, 3) {

            override fun migrate(database: SupportSQLiteDatabase) {
                // Since we didn't alter the table, there's nothing else to do here.
            }

        }

        private val MIGRATION_4_5: Migration = object: Migration(4, 5) {

            override fun migrate(database: SupportSQLiteDatabase) {
                // Since we didn't alter the table, there's nothing else to do here.
            }

        }

    }

}
